### PR TITLE
Fix product id, quantity and price mapping

### PR DESCRIPTION
**Pull request by esadek**
## Description of the change

> Update how we are mapping `product_id`, `quantity` and `item_price` for products in the `products` array.

## Type of change
- [ ] Bug fix (non-breaking change that fixes an issue)
- [ ] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Related issues

> Fix [#1]() 

### Development

- [ ] Lint rules pass locally
- [ ] The code changed/added as part of this pull request has been covered with tests
- [ ] All tests related to the changed code pass in development

### Code review 

- [ ]  This pull request has a descriptive title and information useful to a reviewer. There may be a screenshot or screencast attached
- [ ] "Ready for review" label attached to the PR and reviewers mentioned in a comment
- [ ] Changes have been reviewed by at least one other engineer
- [ ] Issue from task tracker has a link to this pull request

---
This change is [Reviewable](https://reviewable.io/reviews/rudderlabs/rudder-sdk-js/502)

### CODE REVIEW

**Review by shrouti1507 — CHANGES_REQUESTED** [review]
`quantity: products[i].quantity || 1` at line no 352

File: integrations/FacebookPixel/browser.js [hunk 1]
```diff
@@ -344,15 +345,14 @@ class FacebookPixel {
       const contents = [];
       if (products) {
         for (let i = 0; i < products.length; i++) {
-          const pId = products[i].product_id;
+          const pId = products[i].product_id || products[i].sku || products[i].id;
           contentIds.push(pId);
           const content = {
             id: pId,
-            quantity,
+            quantity: products[i].quantity,
```
[the remark this answers is not in the document]
Comment (saikumarrs) [reply integrations/FacebookPixel/browser.js:352]: Once CS gives us a head start, we are supposed to own and drive them to completion. They wouldn't necessarily have enough time to thoroughly implement and test the changes.
CC @utsabc

[the remark this answers is not in the document]
Comment (utsabc) [reply integrations/FacebookPixel/browser.js:352]: Yes @shrouti1507 we are owning this lets drive the changes internally

**Comment by utsabc** [comment]
@shrouti1507 once you make the necessary changes we will plane release with CS team support

File: integrations/FacebookPixel/browser.js [hunk 2]
```diff
@@ -168,6 +168,9 @@ class FacebookPixel {
       if (products && Array.isArray(products)) {
         products.forEach((product) => {
           const productId = product.product_id || product.sku || product.id;
+          if (!isDefined(productId)) {
+            logger.error ("Product id is required. Event not sent");
+          }
```
Comment (utsabc) [inline integrations/FacebookPixel/browser.js:173]:
```suggestion
          if (!isDefined(productId)) {
            logger.error ("Product id is required. Event not sent");
            return;
          }
```

File: integrations/FacebookPixel/browser.js [hunk 3]
```diff
@@ -272,6 +278,9 @@ class FacebookPixel {
         }
       }, legacyTo);
     } else if (event === "Product Added") {
+      if(!isDefinedAndNotNullAndNotEmpty(prodId)) {
+        logger.error("Product id is required. Event not sent");
+      }
```
Comment (utsabc) [inline integrations/FacebookPixel/browser.js:283]:
```suggestion
      if(!isDefinedAndNotNullAndNotEmpty(prodId)) {
        logger.error("Product id is required. Event not sent");
        return;
      }
```

File: integrations/FacebookPixel/browser.js [hunk 4]
```diff
@@ -351,9 +360,12 @@ class FacebookPixel {
           contentIds.push(pId);
           const content = {
             id: pId,
-            quantity: products[i].quantity,
+            quantity: products[i].quantity || 1,
             item_price: products[i].price
           };
+          if (!isDefined(content.id)) {
+            logger.error("Product id is required. Event not sent")
+          }
```
Comment (utsabc) [inline integrations/FacebookPixel/browser.js:368]:
```suggestion
          if (!isDefined(content.id)) {
            logger.error("Product id is required. Event not sent")
            return;
          }
```

File: integrations/FacebookPixel/browser.js [hunk 5]
```diff
@@ -442,9 +454,12 @@ class FacebookPixel {
           contentIds.push(pId);
           const content = {
             id: pId,
-            quantity: product.quantity,
+            quantity: product.quantity || 1,
             item_price: product.price,
           };
+          if (!isDefined(content.id)) {
+            logger.error("Product id is required. Event not sent")
+          }
```
Comment (utsabc) [inline integrations/FacebookPixel/browser.js:462]:
```suggestion
          if (!isDefined(content.id)) {
            logger.error("Product id is required. Event not sent")
            return;
          }
```

**Review by utsabc — APPROVED** [review]

**Review by utsabc — APPROVED** [review]

File: integrations/FacebookPixel/browser.js [hunk 6]
```diff
@@ -346,14 +364,19 @@ class FacebookPixel {
...
       if (products) {
         for (let i = 0; i < products.length; i++) {
-          const pId = products[i].product_id;
+          const pId =
+            products[i].product_id || products[i].sku || products[i].id;
           contentIds.push(pId);
           const content = {
             id: pId,
-            quantity,
+            quantity: products[i].quantity || quantity || 1,
```
Comment (utsabc) [inline integrations/FacebookPixel/browser.js:372]: We should use `get` in places like this for js sdk

File: integrations/FacebookPixel/browser.js [hunk 7]
```diff
@@ -438,16 +461,18 @@ class FacebookPixel {
...
         for (let i = 0; i < products.length; i++) {
           const product = products[i];
-          const pId = product.product_id;
+          const pId = product.product_id || product.sku || product.id;
           contentIds.push(pId);
           const content = {
             id: pId,
-            quantity,
-            item_price: price,
+            quantity: product.quantity || properties.quantity || 1,
```
Comment (utsabc) [inline integrations/FacebookPixel/browser.js:468]: Please use `get`

File: integrations/FacebookPixel/browser.js [hunk 8]
```diff
@@ -166,13 +170,19 @@ class FacebookPixel {
...
+            logger.error(
+              `Product id is required for product ${index}. Event not sent`
+            );
+          }
           if (productId) {
             contentIds.push(productId);
             contents.push({
               id: productId,
-              quantity: quantity || 1,
+              quantity: product.quantity || quantity || 1,
```
Comment (utsabc) [inline integrations/FacebookPixel/browser.js:184]: Please use `get`

File: integrations/FacebookPixel/browser.js [hunk 9]
```diff
@@ -101,6 +106,9 @@ class FacebookPixel {
         ];
         // this construcPayload will help to map the traits in the same way as cloud mode
         payload = constructPayload(rudderElement.message, traitsMapper);
+        if (payload.external_id) {
+          payload.external_id = sha256(payload.external_id);
```
Comment (utsabc) [inline integrations/FacebookPixel/browser.js:110]: Should we add backward compatibility for this? @saikumarrs @ItsSudip

**Review by utsabc — APPROVED** [review]

File: integrations/FacebookPixel/browser.js (hunk 9 above)
Comment (ItsSudip) [reply integrations/FacebookPixel/browser.js:110]: I don't think we need backward compatibility here.

**Review by ItsSudip — DISMISSED** [review]

File: integrations/FacebookPixel/browser.js [hunk 10]
```diff
@@ -149,14 +157,18 @@ class FacebookPixel {
...
-      quantity = properties.quantity;
-      category = properties.category;
-      prodId = properties.product_id || properties.id || properties.sku || "";
-      prodName = properties.product_name;
-      value = properties.value;
-      price = properties.price;
-      query = properties.query;
+      products = get(properties, "products");
+      quantity = get(properties, "quantity");
+      category = get(properties, "category");
```
Comment (saikumarrs) [inline integrations/FacebookPixel/browser.js:162]: There is no need to use `get` for immediate children.

File: integrations/FacebookPixel/browser.js [hunk 11]
```diff
@@ -166,13 +178,22 @@ class FacebookPixel {
...
-        products.forEach((product) => {
-          const productId = product.product_id;
+        products.forEach((product, index) => {
+          const productId =
+            get(product, "product_id") ||
+            get(product, "sku") ||
+            get(product, "id");
+          if (!isDefined(productId)) {
+            logger.error(
+              `Product id is required for product ${index}. Event not sent`
```
Comment (saikumarrs) [inline integrations/FacebookPixel/browser.js:188]: This error message is misleading. The event is still sent but the product is skipped.

File: integrations/FacebookPixel/browser.js [hunk 12]
```diff
@@ -223,6 +244,10 @@ class FacebookPixel {
         }
       }, legacyTo);
     } else if (event === "Product Viewed") {
+      if (!isDefinedAndNotNullAndNotEmpty(prodId)) {
+        // not adding index, as only one product is supposed to be present here
+        logger.error("Product id is required. Event not sent");
```
Comment (saikumarrs) [inline integrations/FacebookPixel/browser.js:249]: The event is still sent in the next statement. Either the error message or the implementation is incorrect.

File: integrations/FacebookPixel/browser.js [hunk 13]
```diff
@@ -271,6 +296,10 @@ class FacebookPixel {
         }
       }, legacyTo);
     } else if (event === "Product Added") {
+      if (!isDefinedAndNotNullAndNotEmpty(prodId)) {
+        // not adding index, as only one product is supposed to be present here
+        logger.error("Product id is required. Event not sent");
```
Comment (saikumarrs) [inline integrations/FacebookPixel/browser.js:301]: Same as above.

File: integrations/FacebookPixel/browser.js [hunk 14]
```diff
@@ -438,16 +478,22 @@ class FacebookPixel {
       if (products) {
         for (let i = 0; i < products.length; i++) {
           const product = products[i];
-          const pId = product.product_id;
+          const pId =
+            get(product, "product_id") ||
+            get(product, "sku") ||
+            get(product, "id");
```
Comment (saikumarrs) [inline integrations/FacebookPixel/browser.js:484]: `get` methods not required here but make sure to check for `product`.

File: integrations/FacebookPixel/browser.js [hunk 15]
```diff
@@ -120,6 +128,10 @@ class FacebookPixel {
     if (properties) {
       const { revenue, currency } = properties;
       revValue = this.formatRevenue(revenue);
+      if (!isDefined(revValue)) {
+        logger.error("Revenue could not be converted to number");
```
Comment (saikumarrs) [inline integrations/FacebookPixel/browser.js:132]:
```suggestion
        logger.error("'properties.revenue' could not be converted to a number");
```

File: integrations/FacebookPixel/browser.js [hunk 16]
```diff
@@ -166,16 +178,26 @@ class FacebookPixel {
...
-              id: productId,
-              quantity: quantity || 1,
-            });
+        for (let i = 0; i < products.length; i++) {
+          const product = products[i];
+          if (product) {
+            const productId = product.product_id || product.sku || product.id;
+            if (!isDefined(productId)) {
+              logger.error(
+                `Product id is required for product ${i}. Event not sent`
```
Comment (saikumarrs) [inline integrations/FacebookPixel/browser.js:187]:
```suggestion
                `Product ID is missing for product ${i}`
```

File: integrations/FacebookPixel/browser.js [hunk 17]
```diff
@@ -166,16 +178,26 @@ class FacebookPixel {
...
+              );
+              return;
+            }
+            if (productId) {
```
Comment (saikumarrs) [inline integrations/FacebookPixel/browser.js:191]: This check is not required anymore. It's better to convert it to an else block.

File: integrations/FacebookPixel/browser.js [hunk 18]
```diff
@@ -223,6 +245,11 @@ class FacebookPixel {
         }
       }, legacyTo);
     } else if (event === "Product Viewed") {
+      if (!isDefinedAndNotNullAndNotEmpty(prodId)) {
+        // not adding index, as only one product is supposed to be present here
+        logger.error("Product id is required. Event not sent");
```
Comment (saikumarrs) [inline integrations/FacebookPixel/browser.js:250]:
```suggestion
        logger.error("Product ID is required. Event is dropped.");
```

File: integrations/FacebookPixel/browser.js [hunk 19]
```diff
@@ -271,6 +298,11 @@ class FacebookPixel {
         }
       }, legacyTo);
     } else if (event === "Product Added") {
+      if (!isDefinedAndNotNullAndNotEmpty(prodId)) {
+        // not adding index, as only one product is supposed to be present here
+        logger.error("Product id is required. Event not sent");
```
Comment (saikumarrs) [inline integrations/FacebookPixel/browser.js:303]:
```suggestion
        logger.error("Product ID is required. Event dropped.");
```

File: integrations/FacebookPixel/browser.js [hunk 20]
```diff
@@ -346,16 +378,23 @@ class FacebookPixel {
...
-
-          content.item_price = price;
-
-          contents.push(content);
+          const product = products[i];
+          if (product) {
+            const pId = product.product_id || product.sku || product.id;
+            if (!isDefined(pId)) {
+              logger.error(
+                `Product id is required for product ${i}. Event not sent`
```
Comment (saikumarrs) [inline integrations/FacebookPixel/browser.js:386]:
```suggestion
                `Product ID is missing for product ${i}. Event dropped.`
```

File: integrations/FacebookPixel/browser.js [hunk 21]
```diff
@@ -438,17 +477,22 @@ class FacebookPixel {
...
-          };
-
-          content.item_price = price;
-
-          contents.push(content);
+          if (product) {
+            const pId = product.product_id || product.sku || product.id;
+            if (!isDefined(pId)) {
+              logger.error(
+                `Product id is required for product ${i}. Event not sent`
```
Comment (saikumarrs) [inline integrations/FacebookPixel/browser.js:484]:
```suggestion
                `Product ID is missing for product ${i}. Event dropped.`
```

File: integrations/FacebookPixel/browser.js [hunk 22]
```diff
@@ -120,6 +128,10 @@ class FacebookPixel {
     if (properties) {
       const { revenue, currency } = properties;
       revValue = this.formatRevenue(revenue);
+      if (!isDefined(revValue)) {
```
Comment (saikumarrs) [inline integrations/FacebookPixel/browser.js:131]: @utsabc This method has become too big and risky to maintain. Moreover, some snippets are repeated.
I observed at least two mistakes during this review, which I attribute to the above.

File: integrations/FacebookPixel/browser.js (hunk 17 above)
Comment (saikumarrs) [reply integrations/FacebookPixel/browser.js]: @shrouti1507?

**Review by utsabc — APPROVED** [review]

**Review by saikumarrs — APPROVED** [review]

**Pull request merged**